File: jeffrey-local/pages-local/src/services/api/model/JvmFlag.ts
/*
 * Jeffrey
 * Copyright (C) 2025 Petr Bouda
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Represents a single flag value change with timestamp.
 */
export interface FlagValueChange {
  /** The flag value at this point in time */
  value: string;
  /** When this value was recorded (ISO format) */
  timestamp: string;
}

/**
 * Represents a JVM flag with its current value and change history.
 */
export default interface JvmFlag {
  /** Flag name (e.g., "UseG1GC", "MaxHeapSize") */
  name: string;
  /** Current (latest) flag value */
  value: string;
  /** Flag type: Boolean, Int, UnsignedInt, Long, String */
  type: string;
  /** How the flag was set: Default, Ergonomic, Command line, Management */
  origin: string;
  /** Previous values if the flag changed during recording */
  previousValues: string[];
  /** Whether the flag value changed during the recording */
  hasChanged: boolean;
  /** Optional description of the flag from OpenJDK documentation */
  description?: string;
  /** Chronological list of value changes (latest first), only populated if hasChanged is true */
  changeHistory?: FlagValueChange[];
}

File: jeffrey-microscope/profiles/profile-persistence-api/src/main/java/cafe/jeffrey/provider/profile/api/JvmFlag.java
/*
 * Jeffrey
 * Copyright (C) 2025 Petr Bouda
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cafe.jeffrey.provider.profile.api;

/**
 * Represents a JVM flag extracted from JFR flag events.
 *
 * @param name   Flag name (e.g., "UseStringDeduplication")
 * @param value  Flag value as string
 * @param type   Flag type (Boolean, Int, UnsignedInt, etc.) derived from event type
 * @param origin How the flag was set (Default, Ergonomic, Command line, Management)
 */
public record JvmFlag(
        String name,
        String value,
        String type,
        String origin
) {
}
